File: web/src/types/mcp.ts
import {z} from 'zod';

export const ToolSchema = z
  .object({
    name: z.string(),
    description: z.optional(z.string()),
    inputSchema: z
      .object({
        type: z.literal("object"),
        properties: z.optional(z.object({}).passthrough()),
      })
      .passthrough(),
  })
  .passthrough();

export type Tool = z.infer<typeof ToolSchema>;

export const ListToolsResultSchema = z.object({
  tools: z.array(ToolSchema),
});

export type ListToolsResult = z.infer<typeof ListToolsResultSchema>;

export interface MCPConfigVersion {
  version: number;
  created_by: string;
  created_at: string;
  action_type: 'Create' | 'Update' | 'Delete' | 'Revert';
  name: string;
  tenant: string;
  routers: string;
  servers: string;
  tools: string;
  mcp_servers: string;
  is_active: boolean;
  hash: string;
}

export interface MCPConfigVersionListResponse {
  data: MCPConfigVersion[];
}

// MCP Capabilities types
export const PromptSchema = z.object({
  name: z.string(),
  description: z.optional(z.string()),
  arguments: z.optional(z.array(z.object({
    name: z.string(),
    description: z.optional(z.string()),
    required: z.optional(z.boolean())
  })))
}).passthrough();

export type Prompt = z.infer<typeof PromptSchema>;

export const ResourceSchema = z.object({
  uri: z.string(),
  name: z.string(),
  description: z.optional(z.string()),
  mimeType: z.optional(z.string())
}).passthrough();

export type Resource = z.infer<typeof ResourceSchema>;

export const ResourceTemplateSchema = z.object({
  uriTemplate: z.string(),
  name: z.string(),
  description: z.optional(z.string()),
  mimeType: z.optional(z.string())
}).passthrough();

export type ResourceTemplate = z.infer<typeof ResourceTemplateSchema>;

export interface MCPCapabilities {
  tools?: Tool[];
  prompts?: Prompt[];
  resources?: Resource[];
  resourceTemplates?: ResourceTemplate[];
}

export type CapabilityType = 'tools' | 'prompts' | 'resources' | 'resourceTemplates';

export interface CapabilityItem {
  name: string;
  description?: string;
  type: CapabilityType;
  [key: string]: unknown;
}

export interface CapabilitiesState {
  loading: boolean;
  error: string | null;
  data: MCPCapabilities | null;
  filteredData: MCPCapabilities | null;
  searchTerm: string;
  selectedType: CapabilityType | 'all';
}

File: internal/common/config/mcp.go
package config

import (
	"time"

	"github.com/ifuryst/lol"

	"github.com/amoylab/unla/internal/common/cnst"
	"github.com/amoylab/unla/pkg/mcp"
)

type (
	// MCPServer represents the MCP server data structure
	MCPServer struct {
		Name      string    `json:"name" yaml:"name"`
		Content   MCPConfig `json:"content" yaml:"content"`
		CreatedAt time.Time `json:"createdAt" yaml:"createdAt"`
		UpdatedAt time.Time `json:"updatedAt" yaml:"updatedAt"`
	}

	MCPConfig struct {
		Name       string            `json:"name" yaml:"name"`
		Tenant     string            `json:"tenant" yaml:"tenant"`
		CreatedAt  time.Time         `json:"createdAt" yaml:"createdAt"`
		UpdatedAt  time.Time         `json:"updatedAt" yaml:"updatedAt"`
		DeletedAt  time.Time         `json:"deletedAt,omitempty" yaml:"deletedAt,omitempty"` // non-zero indicates that all information has been deleted
		Routers    []RouterConfig    `json:"routers,omitempty" yaml:"routers,omitempty"`
		Servers    []ServerConfig    `json:"servers,omitempty" yaml:"servers,omitempty"`
		Tools      []ToolConfig      `json:"tools,omitempty" yaml:"tools,omitempty"`
		McpServers []MCPServerConfig `json:"mcpServers,omitempty" yaml:"mcpServers,omitempty"` // proxy mcp servers
	}

	RouterConfig struct {
		Server    string      `json:"server" yaml:"server"`
		Prefix    string      `json:"prefix" yaml:"prefix"`
		SSEPrefix string      `json:"ssePrefix" yaml:"ssePrefix"`
		CORS      *CORSConfig `json:"cors,omitempty" yaml:"cors,omitempty"`
		Auth      *Auth       `json:"auth,omitempty" yaml:"auth,omitempty"`
	}

	CORSConfig struct {
		AllowOrigins     []string `json:"allowOrigins,omitempty" yaml:"allowOrigins,omitempty"`
		AllowMethods     []string `json:"allowMethods,omitempty" yaml:"allowMethods,omitempty"`
		AllowHeaders     []string `json:"allowHeaders,omitempty" yaml:"allowHeaders,omitempty"`
		ExposeHeaders    []string `json:"exposeHeaders,omitempty" yaml:"exposeHeaders,omitempty"`
		AllowCredentials bool     `json:"allowCredentials" yaml:"allowCredentials"`
	}

	ProxyConfig struct {
		Host string `json:"host" yaml:"host"`
		Port int    `json:"port" yaml:"port"`
		Type string `json:"type" yaml:"type"` // http, https, socks5
	}

	ServerConfig struct {
		Name         string            `json:"name" yaml:"name"`
		Description  string            `json:"description" yaml:"description"`
		AllowedTools []string          `json:"allowedTools,omitempty" yaml:"allowedTools,omitempty"`
		Config       map[string]string `json:"config,omitempty" yaml:"config,omitempty"`
	}

	ToolConfig struct {
		Name         string            `json:"name" yaml:"name"`
		Description  string            `json:"description,omitempty" yaml:"description,omitempty"`
		Method       string            `json:"method" yaml:"method"`
		Endpoint     string            `json:"endpoint" yaml:"endpoint"`
		Proxy        *ProxyConfig      `json:"proxy,omitempty" yaml:"proxy,omitempty"`
		Headers      map[string]string `json:"headers,omitempty" yaml:"headers,omitempty"`
		Args         []ArgConfig       `json:"args,omitempty" yaml:"args,omitempty"`
		RequestBody  string            `json:"requestBody"  yaml:"requestBody"`
		ResponseBody string            `json:"responseBody" yaml:"responseBody"`
		InputSchema  map[string]any    `json:"inputSchema,omitempty" yaml:"inputSchema,omitempty"`
	}

	MCPServerConfig struct {
		Type         string                `json:"type" yaml:"type"`                           // sse, stdio and streamable-http
		Name         string                `json:"name" yaml:"name"`                           // server name
		Command      string                `json:"command,omitempty" yaml:"command,omitempty"` // for stdio
		Args         []string              `json:"args,omitempty" yaml:"args,omitempty"`       // for stdio
		Env          map[string]string     `json:"env,omitempty" yaml:"env,omitempty"`         // for stdio
		URL          string                `json:"url,omitempty" yaml:"url,omitempty"`         // for sse and streamable-http
		Policy       cnst.MCPStartupPolicy `json:"policy" yaml:"policy"`                       // onStart or onDemand
		Preinstalled bool                  `json:"preinstalled" yaml:"preinstalled"`           // whether to install this MCP server when mcp-gateway starts
	}

	ArgConfig struct {
		Name        string      `json:"name" yaml:"name"`
		Position    string      `json:"position" yaml:"position"` // header, query, path, body
		Required    bool        `json:"required" yaml:"required"`
		Type        string      `json:"type" yaml:"type"`
		Description string      `json:"description" yaml:"description"`
		Default     string      `json:"default" yaml:"default"`
		Items       ItemsConfig `json:"items,omitempty" yaml:"items,omitempty"`
	}

	ItemsConfig struct {
		Type       string         `json:"type" yaml:"type"`
		Enum       []string       `json:"enum,omitempty" yaml:"enum,omitempty"`
		Properties map[string]any `json:"properties,omitempty" yaml:"properties,omitempty"`
	}

	// MCPConfigVersion represents a version of an MCP configuration
	MCPConfigVersion struct {
		Version    int             `json:"version" yaml:"version"`
		CreatedBy  string          `json:"created_by" yaml:"created_by"`
		CreatedAt  time.Time       `json:"created_at" yaml:"created_at"`
		ActionType cnst.ActionType `json:"action_type" yaml:"action_type"` // Create, Update, Delete, Revert
		Name       string          `json:"name" yaml:"name"`
		Tenant     string          `json:"tenant" yaml:"tenant"`
		Routers    string          `json:"routers" yaml:"routers"`
		Servers    string          `json:"servers" yaml:"servers"`
		Tools      string          `json:"tools" yaml:"tools"`
		McpServers string          `json:"mcp_servers" yaml:"mcp_servers"`
		IsActive   bool            `json:"is_active" yaml:"is_active"` // indicates if this version is currently active
		Hash       string          `json:"hash" yaml:"hash"`           // hash of the configuration content
	}

	// Auth represents authentication configuration
	Auth struct {
		Mode cnst.AuthMode `json:"mode" yaml:"mode"`
	}
)

// ToToolSchema converts a ToolConfig to a ToolSchema
func (t *ToolConfig) ToToolSchema() mcp.ToolSchema {
	// Create properties map for input schema
	properties := make(map[string]any)
	required := make([]string, 0)
	for _, arg := range t.Args {
		property := map[string]any{
			"type":        arg.Type,
			"description": arg.Description,
		}

		if arg.Type == "array" {
			items := make(map[string]any)
			if len(arg.Items.Enum) > 0 {
				items["enum"] = lol.Union(arg.Items.Enum)
			} else {
				items["type"] = arg.Items.Type
				// If items is an object type, recursively process its properties
				if arg.Items.Type == "object" && arg.Items.Properties != nil {
					items["properties"] = arg.Items.Properties
				}
			}
			property["items"] = items
		}

		properties[arg.Name] = property
		if arg.Required {
			required = append(required, arg.Name)
		}
	}

	// Merge with existing input schema if any
	if t.InputSchema != nil {
		for k, v := range t.InputSchema {
			properties[k] = v
		}
	}

	return mcp.ToolSchema{
		Name:        t.Name,
		Description: t.Description,
		InputSchema: mcp.ToolInputSchema{
			Type:       "object",
			Properties: properties,
			Required:   required,
		},
	}
}
